columnflow/calibration/cms/tau.py

```diff
@@ -0,0 +1,259 @@
+    # function to determine the tec config
+    get_tec_config=(lambda self: TECConfig.new(self.config_inst.x.tec)),
+)
+def tec(
+    self: Calibrator,
+    events: ak.Array,
+    **kwargs,
+) -> ak.Array:
+    """
+    Calibrator for tau energy. Requires an external file in the config under ``tau_sf``, e.g.
+
+    .. code-block:: python
+
+        cfg.x.external_files = DotDict.wrap({
+            "tau_sf": "/afs/cern.ch/work/m/mrieger/public/mirrors/jsonpog-integration-6ce37404/POG/TAU/2017_UL/tau.json.gz",  # noqa
+        })
+
+    and a :py:class:`TECConfig` configuration object named ``tec``,
+
+    .. code-block:: python
+
+        # run 3 example
+        from columnflow.calibration.cms.tau import TECConfig
+
+        cfg.x.tec = TECConfig(
+            tagger="DeepTau2018v2p5",
+            corrector_kwargs={"wp": "Tight", "wp_VSe": "Tight"},
+        )
+
+    *get_tau_file* and *get_tec_config* can be adapted in a subclass in case they are stored
+    differently in the config.
+
+    Resources:
+    https://twiki.cern.ch/twiki/bin/view/CMS/TauIDRecommendationForRun2?rev=113
+    https://gitlab.cern.ch/cms-nanoAOD/jsonpog-integration/-/blob/849c6a6efef907f4033715d52290d1a661b7e8f9/POG/TAU
+    """
+    # fail when running on data
+    if self.dataset_inst.is_data:
+        raise ValueError("attempt to apply tau energy corrections in data")
+
+    # the correction tool only supports flat arrays, so convert inputs to flat np view first
+    pt = flat_np_view(events.Tau.pt, axis=1)
+    mass = flat_np_view(events.Tau.mass, axis=1)
+    eta = flat_np_view(events.Tau.eta, axis=1)
+    dm = flat_np_view(events.Tau.decayMode, axis=1)
+    match = flat_np_view(events.Tau.genPartFlav, axis=1)
+
+    # get the scale factors for the four supported decay modes
+    dm_mask = (dm == 0) | (dm == 1) | (dm == 10) | (dm == 11)
+
+    # prepare arguments
+    variable_map = {
+        "pt": pt[dm_mask],
+        "eta": eta[dm_mask],
+        "dm": dm[dm_mask],
+        "genmatch": match[dm_mask],
+        "id": self.tec_config.tagger,
+        **self.tec_config.corrector_kwargs,
+    }
+    args = tuple(
+        variable_map[inp.name] for inp in self.tec_corrector.inputs
+        if inp.name in variable_map
+    )
+
+    # nominal correction
+    scales_nom = np.ones_like(dm_mask, dtype=np.float32)
+    scales_nom[dm_mask] = self.tec_corrector(*args, "nom")
+
+    # varied corrections
+    if self.with_uncertainties:
+        scales_up = np.ones_like(dm_mask, dtype=np.float32)
+        scales_up[dm_mask] = self.tec_corrector(*args, "up")
+        scales_down = np.ones_like(dm_mask, dtype=np.float32)
+        scales_down[dm_mask] = self.tec_corrector(*args, "down")
+
+    # custom adjustment 1: reset where the matching value is unhandled
+    # custom adjustment 2: reset electrons faking taus where the pt is too small
+    mask1 = (match < 1) | (match > 5)
+    mask2 = ((match == 1) | (match == 3) & (pt <= 20.0))
```

Comment:
Not sure about the details, but this currently puts the SF of all taus with match == 1 to 1.0
```suggestion
    mask2 = ((match == 1) | (match == 3)) & (pt <= 20.0)
```

Reply:
Good catch :+1: